aptos-move/framework/aptos-stdlib/sources/cryptography/algebra.move

```diff
@@ -0,0 +1,69 @@
+/// This module provides generic structs/functions for operations of algebraic structures (e.g. fields and groups),
+/// which can be used to build generic cryptographic schemes atop.
+/// See `algebra_*.move` for currently implemented algebraic structures.
+///
+/// Below are the operations currently supported.
+/// - Element serialization/deserialization.
+/// - Field operations.
+///   - Addition.
+module aptos_std::algebra {
+    use std::option::{Option, some, none};
+    use std::features::generic_algebraic_structures_basic_operations_enabled;
+
+    /// This struct represents an element of an algebraic structure `S`.
+    struct Element<phantom S> has copy, drop {
+        handle: u64
+    }
+
+    // Public functions begin.
+
+    /// Compute `x + y` for elements `x` and `y` of a field `S`.
+    public fun field_add<S>(x: &Element<S>, y: &Element<S>): Element<S> {
```

Comment:
So, I don't think it makes sense to have both 
(1) a `field_add` function here for `Element<bls12381::Fr>`'s, and
(2) later on, also have a `group_add` for, say,`Element<bls12381::G1>`'s

These should be the same function called `add`, no?

This applies to many other functions in your other, larger `algebra.move` PR.

For operations that are only supported on groups/fields, we can have a `group_`/`field_` prefix. But many of them are shared.

Reply:
I'm leaning to keep `group_add` and `field_add` separate because they refer to the operation in a group, and one of the 2 operations in a field respectively, which are from 2 definitions.

Though I have a generic function `eq<S>()` for any generic structure, since this refers to the element equality check universally defined in every set.

Reply:
But such separation would not be consistent with the spirit of this module, which is to treat every "algebraic structure" as generically as possible and abort if some functionality is not possible for that structure, no?

Especially if we are inconsistent about treating some things generically but not others: `field_add` but not `field_eq` (instead `eq`). 

To me, it seems like `add, sub, neg, eq` should be shared between groups and fields (assuming we're treating groups additively).

Plus, we could save an extra native for `field_add` too (and many others).

Plus, we could keep our module smaller.

Plus, remember we can always add inline functions in `algebra_bls12381` for all the `field_*` functions we want.

Reply:
In the end, if Move supported binary operations, we would do `scalar1 + scalar2` just like we do `group_el1 + group_el2`, it seems like having a single `add<S>` is just a step in that direction.

Reply:
I think the spirit of this module is, for example, to generalize `bls12381g1_pairing()`, `bn254_pairing()` etc. into generic `pairing<>()` so one can write a generic groth16 verifier.  Combining `field_add()` and `group_add()` has no such benefit.

`eq()` is special, and actually not referring to `field_eq` or `group_eq`. It refers to the element equality check operation that is implicitly but universally defined whenever we work with a set. It applies to every algebraic structure I'm aware of. That said, I think having `field_eq` and `group_eq` is also fine.

If we combine `group_add()` and `field_add()` into `add()`, we get a function with its meaning subject to change. I'm not sure how to documentate such `add()`... "Compute x, y where x, y are from an addable structure S"? I'm feeling nervous writing this down because in our case, it's an extra decision to use additive notation for groups. What if it turns out additive notation is a bad idea? I actually still think `group_op` might be less ambiguous and therefore might be a safer name, and `group_add` might give the impression that this op only applies to abelian groups.

I also like to have fewer native functions, but I'd remove those derivable ones, for example `group_sub()`, which can be derived by `group_neg()` and `group_add()`.

Reply:
We agreed asynchronously that:
1. `algebra.move` is the equivalent of an interface for any algebraic structure (currently with support for monoids, groups, fields, and rings)
2. We can give examples of how to use `algebra.move` inside a, say, `field_algebra.move` module, which would just have inline functions that forward calls to the relevant functions in `algebra.move`. We could do the same for `group.move`.
3. Overall, our goal here is to keep this module generic enough so that we can build whatever we may need in the future on top of it (e.g., rings, or lattices)